File: learn_ai/MCMC/Monte-Carlo/Monte-Carlo.py
import random
import math
import argparse
from statistics import mean, stdev

def monte_carlo_simulation(square_size: int, num_samples: int, iterations: int = 5):
    """
    蒙特卡洛模拟估算圆面积
    :param square_size: 正方形边长
    :param num_samples: 每次迭代采样数
    :param iterations: 迭代次数 (默认5次)
    :return: 统计结果字典
    """
    radius = square_size / 2.0
    real_area = math.pi * radius ** 2
    square_area = square_size ** 2
    
    predictions = []
    
    for iter in range(iterations):
        inside = 0
        
        # 生成随机点并统计
        for _ in range(num_samples):
            x = random.uniform(0, square_size)
            y = random.uniform(0, square_size)
            
            # 计算点与圆心的距离
            dx = x - radius
            dy = y - radius
            if math.hypot(dx, dy) <= radius:
                inside += 1
                
        # 计算预测面积
        predicted = (inside / num_samples) * square_area
        predictions.append(predicted)
        
        # 打印单次结果
        print(f"Iter {iter+1:2} => Predicted: {predicted:.2f} | Error: {real_area - predicted:+.2f}")

    # 计算统计指标
    avg_pred = mean(predictions)
    std_pred = stdev(predictions) if len(predictions) > 1 else 0.0
    min_pred = min(predictions)
    max_pred = max(predictions)
    
    return {
        "real_area": real_area,
        "average": avg_pred,
        "std_dev": std_pred,
        "min": min_pred,
        "max": max_pred,
        "samples": num_samples,
        "iterations": iterations
    }

if __name__ == "__main__":
    # 配置命令行参数解析
    parser = argparse.ArgumentParser(description="蒙特卡洛模拟估算圆面积")
    parser.add_argument("square", type=int, help="正方形边长")
    parser.add_argument("samples", type=int, help="每次迭代采样数")
    parser.add_argument("-i", "--iterations", type=int, default=5, 
                       help="迭代次数 (默认: 5)")
    
    args = parser.parse_args()
    
    # 参数验证
    if args.square <= 0 or args.samples <= 0 or args.iterations <= 0:
        print("参数必须为正整数")
        exit(1)
    
    # 执行模拟
    results = monte_carlo_simulation(args.square, args.samples, args.iterations)
    
    # 打印最终结果
    print("\n===== 最终结果 =====")
    print(f"理论面积:      {results['real_area']:.2f}")
    print(f"平均预测值:    {results['average']:.2f} (±{results['std_dev']:.2f})")
    print(f"最小值:        {results['min']:.2f}")
    print(f"最大值:        {results['max']:.2f}")
    print(f"迭代次数:      {results['iterations']}")
    print(f"总采样数:      {results['samples'] * results['iterations']:,}")
    print(f"绝对偏差:      {results['real_area'] - results['average']:+.2f}")
    print(f"相对误差:      {abs(results['real_area'] - results['average']) / results['real_area']:.2%}")

File: learn_ai/MCMC/Monte-Carlo/Monte-Carlo.cpp
#include <charconv>
#include <cmath>
#include <fmt/format.h>
#include <random>
#include <vector>
#include <numeric>
using XYpoint = std::pair<double, double>;

double length_line(const XYpoint& point, double circle) {
    double rel_x = point.first - circle;
    double rel_y = point.second - circle;
    return std::hypot(rel_x, rel_y);
}

int main(int argc, char* argv[]) {
    // 参数解析与验证
    if (argc < 3 || argc > 4) {
        fmt::print("Usage: {} <Square> <Samples> [Iterations=5]\n", argv[0]);
        fmt::print("Example: {} 20 100000 10\n", argv[0]);
        return EXIT_FAILURE;
    }

    int square = 0, sample = 0, iterations = 5;
    auto [ptr1, ec1] = std::from_chars(argv[1], argv[1] + strlen(argv[1]), square);
    auto [ptr2, ec2] = std::from_chars(argv[2], argv[2] + strlen(argv[2]), sample);
    
    if (argc == 4) {
        auto [ptr3, ec3] = std::from_chars(argv[3], argv[3] + strlen(argv[3]), iterations);
        if (ec3 != std::errc()) iterations = 5;
    }

    if (ec1 != std::errc() || ec2 != std::errc() || 
        square <= 0 || sample <= 0 || iterations <= 0) {
        fmt::print("Invalid parameters\n");
        return EXIT_FAILURE;
    }

    const double radius = square / 2.0;
    const double real_area = M_PI * radius * radius;
    const double square_area = square * square;
    
    // 存储每次迭代结果
    std::vector<double> predictions;
    predictions.reserve(iterations);

    std::random_device rd;
    
    // 执行多次蒙特卡洛模拟
    for (int iter = 0; iter < iterations; ++iter) {
        std::mt19937 gen(rd());
        std::uniform_real_distribution<double> dist(0.0, square);
        
        int inside = 0;
        for (int i = 0; i < sample; ++i) {
            XYpoint point = {dist(gen), dist(gen)};
            if (length_line(point, radius) <= radius) {
                ++inside;
            }
        }
        
        double predicted = (inside / static_cast<double>(sample)) * square_area;
        predictions.push_back(predicted);
        
        fmt::print("Iter {:2} => Predicted: {:.2f} | Deviation:: {:+.2f}\n", 
                   iter+1, predicted, real_area - predicted);
    }

    // 统计计算结果
    const double mean = std::accumulate(predictions.begin(), predictions.end(), 0.0)
                      / iterations;
    
    double variance = 0.0;
    for (double pred : predictions) {
        variance += std::pow(pred - mean, 2);
    }
    variance /= iterations;
    
    const double std_dev = std::sqrt(variance);

    // 最终输出
    fmt::print("\n===== Final Results =====\n");
    fmt::print("Real area:      {:.4f}\n", real_area);
    fmt::print("Average pred:   {:.4f} (±{:.4f})\n", mean, std_dev);
    fmt::print("Min pred:       {:.4f}\n", *std::min_element(predictions.begin(), predictions.end()));
    fmt::print("Max pred:       {:.4f}\n", *std::max_element(predictions.begin(), predictions.end()));
    fmt::print("Deviation:      {:+.4f}\n", real_area - mean);
    fmt::print("Relative Deviation: {:.4f}\n", std::abs(real_area - mean) / real_area);

    return EXIT_SUCCESS;
}
